presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestFlattenNestedConcat.java

```diff
@@ -0,0 +1,70 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.sql.planner.iterative.rule.test.PlanBuilder;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.sql.planner.iterative.rule.FlattenNestedConcat.flattenExpression;
+import static org.testng.Assert.assertEquals;
+
+public class TestFlattenNestedConcat
+{
+    private static void assertRewritten(String from, String to)
+    {
+        assertEquals(flattenExpression(PlanBuilder.expression(from)), PlanBuilder.expression(to));
+    }
+
+    private static void assertNotRewritten(String expression)
+    {
+        assertEquals(flattenExpression(PlanBuilder.expression(expression)), PlanBuilder.expression(expression));
+    }
+
+    @Test
+    public void testRewriteMapConcatAndArrayConcat()
+    {
+        // nested map_concat
+        assertRewritten(
+                "map_concat(map(ARRAY[1,2], ARRAY['a','b']), map_concat(map(ARRAY[3,4], ARRAY['c','d']), map(ARRAY[5,6], ARRAY['e', 'f'])))",
+                "map_concat(map(ARRAY[1,2], ARRAY['a','b']), map(ARRAY[3,4], ARRAY['c','d']), map(ARRAY[5,6], ARRAY['e', 'f']))");
+
+        assertRewritten(
+                "map_concat(map_concat(map_concat(map,map),map),map,map_concat(map,map))",
+                "map_concat(map,map,map,map,map,map)");
+
+        // nested array_concat
+        assertRewritten(
+                "map(array_concat(array_concat(ARRAY[1]),array_concat(ARRAY[2])), ARRAY[3,4])",
+                "map(array_concat(ARRAY[1],ARRAY[2]), ARRAY[3,4])");
+
+        assertRewritten(
+                "array_concat(array[1,2],array[3,4], array_concat(array[5,6],array_concat(array[7,8])))",
+                "array_concat(array[1,2],array[3,4],array[5,6],array[7,8])");
+
+        // nested map_concat and array_concat
+        assertRewritten(
+                "map_concat(map(array_concat(array_concat(ARRAY[1]),array_concat(ARRAY[2])), ARRAY['a','b']), map_concat(map(ARRAY[3,4], ARRAY['c','d']), map(ARRAY[5,6], ARRAY['e', 'f'])))",
+                "map_concat(map(array_concat(ARRAY[1],ARRAY[2]), ARRAY['a','b']), map(ARRAY[3,4], ARRAY['c','d']), map(ARRAY[5,6], ARRAY['e', 'f']))");
+
+        // assert does not rewrite non direct children functions
+        assertNotRewritten(
+                "map_concat(map, map(map_concat), map)");
+
+        assertNotRewritten(
+                "map_concat(map(map_concat(map,map),map),map,map(map,map))");
+
+        assertNotRewritten(
+                "array_concat(array, array[array_concat], array)");
+    }
```

Comment:
I'm not sure this is enough. Since this is a plan rewrite, you should add tests that test the plans to make sure they have the right shape. Similarly add other some end-to-end tests as well to test that the rewrite preseves the results inncluding some null values etc.